src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3359,48 +3421,55 @@ private bool HasTopLevelNullabilityConversion(TypeSymbolWithAnnotations source,
                     }
                     break;
 
+                case ConversionKind.ImplicitThrow:
+                    break;
+
                 case ConversionKind.Unboxing:
-                    if (!operandType.IsNull && targetType.IsTypeParameter())
+                    if (!operandType.IsNull)
                     {
-                        resultAnnotation = operandType.GetValueNullableAnnotation();
+                        if (targetType.IsTypeParameter())
+                        {
+                            resultAnnotation = operandType.GetValueNullableAnnotation();
 
-                        if (resultAnnotation == NullableAnnotation.NotAnnotated)
+                            if (resultAnnotation == NullableAnnotation.NotAnnotated)
+                            {
+                                resultAnnotation = NullableAnnotation.NotNullable;
+                            }
+                        }
+                        else if (targetType.IsValueType)
                         {
-                            resultAnnotation = NullableAnnotation.NotNullable;
+                            resultAnnotation = (targetType.IsNullableType() && operandType.NullableAnnotation.IsAnyNullable()) ? NullableAnnotation.Nullable : NullableAnnotation.NotNullable;
```

Comment:
>resultAnnotation = (targetType.IsNullableType() && operandType.NullableAnnotation.IsAnyNullable()) ? NullableAnnotation.Nullable : NullableAnnotation.NotNullable; [](start = 28, length = 162)

This doesn't feel right. I think unboxing into a nullable value type can always yield null value (due to the underlying type mismatch) unless we are unboxing from the same nullable type, which could probably happen with type parameters constrained to a nullable value type.

Reply:
Also, I am not sure if ```operandType.IsNull``` can happen when target type is nullable value type, but I would also just set state to nullable for this case.

---
In reply to: [239930911](https://github.com/dotnet/roslyn/pull/31499#discussion_r239930911) [](ancestors = 239930911)

Reply:
Never mind that applies to "as" conversion.

---
In reply to: [239931254](https://github.com/dotnet/roslyn/pull/31499#discussion_r239931254) [](ancestors = 239931254,239930911)